exercises/src/test/scala/datastructures/ListSpec.scala

```diff
@@ -0,0 +1,417 @@
+      List.drop(List(1), 1) mustEqual Nil
+      List.drop(List('a'), 2) mustEqual Nil
+    }
+
+    "remove all elements from any List where length <= n" in {
+      List.drop(List(1,2), 2) mustEqual Nil
+      List.drop(List('a','b'), 4) mustEqual Nil
+    }
+
+    "remove first n elements for multiple-element List" in {
+      List.drop(List(1,2), 1) mustEqual List(2)
+      List.drop(List('a','b','c'), 2) mustEqual List('c')
+    }
+  }
+
+  //========================================================================================
+  //#3.5
+  "dropWhile" should {
+    "result in Nil for an empty List" in {
+      List.dropWhile(List())(x => true) mustEqual Nil
+      List.dropWhile(List[Int]())(x => x>5) mustEqual Nil
+    }
+
+    "remove the only element in a single-element List matching the predicate" in {
+      List.dropWhile(List(1))(_==1) mustEqual Nil
+      List.dropWhile(List('a'))(_<'z') mustEqual Nil
+    }
+
+    "remove all elements in a multiple-element List matching the predicate" in {
+      List.dropWhile(List(1,2,3))(_ < 4) mustEqual Nil
+      List.dropWhile(List('a','b','c'))(_<'z') mustEqual Nil
+    }
+
+    "remove only first elements matching the predicate in a multiple-element List" in {
+      List.dropWhile(List(5,6,1,2,3))(_ > 4) mustEqual List(1,2,3)
+      List.dropWhile(List('a','b','c'))(_=='a') mustEqual List('b','c')
+    }
+
+    "not remove anything in a List with first element not matching the predicate" in {
+      List.dropWhile(List(5,6,1,2,3))(_ < 4) mustEqual List(5,6,1,2,3)
+      List.dropWhile(List('a','b','c'))(_=='b') mustEqual List('a','b','c')
+    }
+  }
+
+  //========================================================================================
+  //#3.6
+  "init" should {
+    "result in Nil for an empty List" in {
+      List.init(List()) mustEqual Nil
+    }
+
+    "remove the only element in a single-element List" in {
+      List.init(List(1)) mustEqual Nil
+      List.init(List('a')) mustEqual Nil
+    }
+
+    "remove the last element in a multiple-element List" in {
+      List.init(List(1,2,3)) mustEqual List(1,2)
+      List.init(List('a','b','c')) mustEqual List('a','b')
+    }
+    //In order to finish the init call the case Cons(_,Nil) => Nil should match. Which means that we have to call init n-1 times where n is the list length
+  }
+
+  //========================================================================================
+  //#3.8
+  "foldRight" should {
+    "result in original list when passed the case class constructors" in {
+      List.foldRight(List(), Nil:List[Int])(Cons(_,_)) mustEqual List()
+      List.foldRight(List(1,2,3), Nil:List[Int])(Cons(_,_)) mustEqual List(1,2,3)
+    }
+  }
+
+  //========================================================================================
+  //#3.9
+  "length" should {
+    "result in 0 for an empty List" in {
+      List.length(List()) mustEqual 0
+    }
+
+    "result in 1 for a single-element List" in {
```

Comment:
Here also you can use property-based testing to good effect (`prop { x: Int => List.length(List(x)) mustEqual 1 }`) instead of having to write multiple cases.